PurchasesCoreSwift/Caching/DeviceCache.swift

```diff
@@ -0,0 +1,491 @@
+//
+//  DeviceCache.swift
+//  PurchasesCoreSwift
+//
+//  Created by Joshua Liebowitz on 7/13/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+import Foundation
+
+// swiftlint:disable file_length
+@objc(RCDeviceCache) open class DeviceCache: NSObject {
+
+    // Thread-safe, but don't call from anywhere inside this class.
+    @objc open var cachedAppUserID: String? { readCache { self.threadUnsafeCachedAppUserID } }
+    @objc open var cachedLegacyAppUserID: String? {
+        return readCache {
+            return self.userDefaults.string(forKey: CacheKeys.legacyGeneratedAppUserDefaults)
+        }
+    }
+    @objc open var cachedOfferings: Offerings? { offeringsCachedObject.cachedInstance() }
+
+    private let cacheDurationInSecondsInForeground = 60 * 5.0
+    private let cacheDurationInSecondsInBackground = 60 * 60 * 25.0
+    private let accessQueue = DispatchQueue(label: "DeviceCacheQueue")
+    private var threadUnsafeCachedAppUserID: String? { userDefaults.string(forKey: CacheKeys.appUserDefaults.rawValue) }
+    private let userDefaults: UserDefaults
+    private let notificationCenter: NotificationCenter
+    private let offeringsCachedObject: InMemoryCachedObject<Offerings>
+    private var appUserIDHasBeenSet: Bool = false
+    private let assertionFunction: (String) -> Void
+
+    @objc convenience public init(userDefaults: UserDefaults = UserDefaults.standard) {
+        self.init(userDefaults: userDefaults, offeringsCachedObject: nil, notificationCenter: nil)
+    }
+
+    public init(userDefaults: UserDefaults = UserDefaults.standard,
+                offeringsCachedObject: InMemoryCachedObject<Offerings>? = InMemoryCachedObject(),
+                notificationCenter: NotificationCenter? = NotificationCenter.default,
+                assertionFunction: @escaping (String) -> Void = { assertionFailure($0) }) {
```

Comment:
why is this needed?

Reply:
In order to test assertions in non-x86_64 platforms, we need to inject the assertion logic so that under normal circumstances, we we just `assertionFailure`, but during tests on ARM64, we just ensure this was called. You can check out the handling: https://github.com/RevenueCat/purchases-ios/pull/658/files#diff-f3a5243addd4ec9aa6a72782c2496bf77defcf852e8d6498ae3232bd88bbeddbR213

`assertionFailures` can't be tested anymore due to CPU architecture specifics 🙃